### ISSUE TITLE
Cant link dll

### ISSUE THREAD

**Issue by catinwarmhands** [Julgodis/Cupcake-Compiler#2]
I have no idea why, but i cant link my test dll file
File `mylib.c`:
```c
#include <stdio.h>
#include <inttypes.h>

typedef  int32_t s32;

#define DLLEXPORT __declspec(dllexport)

DLLEXPORT void my_hello(void) {
	printf("hi from my_hello!\n");
}

DLLEXPORT s32 my_sum(s32 a, s32 b) {
	return a + b;
}
```
File `main.jai`:
```
#import "modules/Print.jai"

#foreign_library "mylib"
my_hello  :: #no_context () -> Void #foreign;
my_sum    :: #no_context (a: s32, b: s32) -> s32 #foreign;

main :: () {
	my_hello();
	print("my_sum test: %", my_sum(24, 4));
}

#run main()
```
Build process:
```
> clang -shared mylib.c -o mylib.dll
#or
> tcc -shared mylib.c

> cupcake main.jai
Cupcake (v0.1-alpha)
* External: 'my_hello' 6 (0)
* External: 'my_sum' 6 (0)
* Function: 'main' 2 (0)
*      Run: 3
-------------------------------------------
Workspace 'program' (2 errors, 2 reports)
.\main.jai:4,1: Error: Unknown foreign procedure 'my_hello' (my_hello)


    >
    > #foreign_library "mylib"
    > my_hello  :: #no_context () -> Void #foreign;
    > ^

.\main.jai:5,1: Error: Unknown foreign procedure 'my_sum' (my_sum)


    > #foreign_library "mylib"
    > my_hello  :: #no_context () -> Void #foreign;
    > my_sum    :: #no_context (a: s32, b: s32) -> s32 #foreign;
    > ^


There were errors, exiting!

```

**Comment by Julgodis** [comment]
It seems that the problem is that Cupcake can't find the correct names in your DLL. The cause is very likely name mangling. Cupcake only support C style name mangling (for now) i.e the type of the function you're importing doesn't matter.

When using MSVC i get the same problem. The solution was to explicitly tell the compiler to use C name mangling by using `extern "C" `.

```c++
#include <stdio.h>
#include <inttypes.h>

typedef  int32_t s32;

#define DLLEXPORT extern "C" __declspec(dllexport)

DLLEXPORT void my_hello(void) {
	printf("hi from my_hello!\n");
}

DLLEXPORT s32 my_sum(s32 a, s32 b) {
	return a + b;
}
```

```
Cupcake (v0.1-alpha)
hi from my_hello!
my_sum test: 28
```

Another solution is to specify the mangled name after the `#foreign` directive.
With MSVC this jai code works without change the DLL C code.

```jai
#import "modules/Print.jai"

#foreign_library "mylib"
my_hello  :: #no_context () -> Void #foreign "?my_hello@@YAXXZ";
my_sum    :: #no_context (a: S32, b: S32) -> S32 #foreign "?my_sum@@YAHHH@Z";

main :: () {
	my_hello();
	print("my_sum test: %", my_sum(24, 4));
}

#run main()
```

I believe for Clang you should use:
```jai
my_hello  :: #no_context () -> Void #foreign "_Z8my_hellov";
my_sum    :: #no_context (a: S32, b: S32) -> S32 #foreign "_Z6my_sumii";
```

The mangled names can be exported from your DLL with `dumpbin /exports mylib.dll` (on Windows).

**Comment by catinwarmhands** [comment]
I checked name mangling via [DLL Export Viewer](http://www.nirsoft.net/utils/dll_export_viewer.html)  and `.def` file from `tcc -impdef` and it seems like functions named as they should be:
![img](https://i.imgur.com/pEy6rDh.png)

I got this result for both `tcc` and `clang` , since they are c (not c++) compilers.
`clang++` with `extern "c"` also produces same result.

I tried to use `clang++`  and specify the mangled name after the `#foreign` directive, but got same error as before.

UPD: It works fine with `gcc` or `g++` with and without `extern "c"`. Thatis strange, but i guess this is fine.
